File: javascript/minswaps.js
function minimumSwaps(arr) {
    let ans = 0;
    const visited = {};
    const startState = arr.map((val, idx) => {
        visited[idx] = false;
        return [idx, val];
    });

    // sort based on arr values
    const endState = startState.sort((a, b) => a[1] - b[1]);

    for (let i = 0; i < arr.length; i++) {
        if (visited[i] || endState[i][0] === i) {
            continue;
        }

        answ += calcCycleSize(i, endState, visited);
    }
    return ans;
}

//traverse graph
function calcCycleSize(i, arr, visited) {
    let cycleSize = 0;
    let j = i;
    // while not full circle
    while (!visited[j]) {
        visited[j] = true;
        j = arr[j][0];
        cycleSize++;
    }
    if (cycleSize > 0) {
        return cycleSize - 1;
    }
    return 0;
}

File: python/min_swaps.py
def minSwaps(arr): 
    n = len(arr) 
    print("arr ", arr)
      
    # Create two arrays and use  
    # as pairs where first array  
    # is element and second array 
    # is position of first element 
    arrpos = [*enumerate(arr)] 
    print("before sorting")
    print(arrpos)
      
    # Sort the array by array element  
    # values to get right position of  
    # every element as the elements  
    # of second array. 
    arrpos.sort(key = lambda it:it[1]) 
    print("after sorting")
    print(arrpos)
      
    # To keep track of visited elements.  
    # Initialize all elements as not  
    # visited or false. 
    vis = {k:False for k in range(n)} 
      
    # Initialize result 
    ans = 0
    for i in range(n): 
        print("i ", i)
          
        # alreadt swapped or  
        # alreadt present at  
        # correct position 
        if vis[i] or arrpos[i][0] == i: 
            continue
              
        # find number of nodes  
        # in this cycle and 
        # add it to ans 
        cycle_size = 0
        j = i 
        while not vis[j]: 
              
            # mark node as visited 
            vis[j] = True

            print("j ",j)
              
            # move to next node 
            j = arrpos[j][0] 

            print("next node ", j)
            cycle_size += 1
            print("cycle size", cycle_size)
              
        # update answer by adding 
        # current cycle 
        if cycle_size > 0: 
            ans += (cycle_size - 1) 
    # return answer 
    return ans 
  
arr = [1, 5, 4, 3, 2] 
print(minSwaps(arr)) 
  